fossas/spectrometer
# Partial Graph Uploads

# Overview

[Tech plan](https://docs.google.com/document/d/1q0TC0jR_wT4rxi2DMhkJA9UYyMNAs98WpaV5tAAKTAM/edit#heading=h.em4xrcdbhsh5)

Right now, FOSSA assumes that all dependencies from `srclib` analyses are direct dependencies, and that all dependencies from the CLI are a full graph, with no in-between. Because of this, we take the results from a `srclib` scan, and analyze each dependency, and then analyze the results from each of those scans for dependencies, and so on and so forth. This is often referred to as the “recursive step”, and is the strategy we use to construct dependency graphs for non-CLI builds. We do this, because we are unable to construct a full graph due to a lack of information provided from the source code.

In order to replace `srclib` with the `CLI`, we will need to add this concept to `CLI` output. This PR introduces the concept of `GraphBreadth` to the graph constructed in each strategy. This will be either `Complete`, or `Partial`, and signal to Core if it needs to do further dependency analysis or not

## Acceptance criteria

- Adds a `GraphBreadth` section to each SourceUnit in the output payload of `analyze`

## Testing plan

- Run analysis on several project types
- Check each strategy and confirm that it is in fact `Complete` or `Partial`

## Risks

- Low risk, as this is an addition to the payload, and will be unused for now

## References

[Tech plan](https://docs.google.com/document/d/1q0TC0jR_wT4rxi2DMhkJA9UYyMNAs98WpaV5tAAKTAM/edit#heading=h.em4xrcdbhsh5)

## Checklist

~- [ ] I added tests for this PR's change (or confirmed tests are not viable).~
~- [ ] If this PR introduced a user-visible change, I added documentation into `docs/`.~
